[ENT-423] Automatically populate Country field for SSO scenarios

UmanShahzad opened the pull request:
There are some use cases in which a required field coming from a SAML IdP is not recognized by Open edX, such as the country code. This PR has a simple aim -- recognize the country code attribute from SAML IdPs and pass it on to the registration flow, where the user API already utilizes the ISO-3166-1 Alpha-2 country code under the key `country`, but will now also recognize it when coming from SAML IdPs.

**JIRA tickets**: Implements [ENT-423](https://openedx.atlassian.net/browse/ENT-423).

**Discussions**: See summary above.

**Dependency**: This is apparently blocked by [ENT-365](https://openedx.atlassian.net/browse/ENT-365), whose PR is up at https://github.com/edx/edx-platform/pull/15135 -- it is stable, has been approved by an OpenCraft developer and just needs confirmation from 1 more third-party-auth domain expert.

**Deployment targets**: edx.org and edge.edx.org

**Merge deadline**: Urgent.

**Sandbox**: TBD

**Testing instructions**:

- Configure SAML such that a country attribute is returned with some URN.
- Input this URN into the SAMLProvider under the Country attribute.
- Attempt login with SAML provider.
- Notice country form field filled with what SAML IdP passed in.

(more specific instructions to come, but if you understand the above then give it a shot).

**Author notes and concerns**: 

1. I've worked on a more *general* version of the same problem -- instead of explicitly having to add attributes the way we do in this PR, in the future it'll all be JSON based and put into a single "attributes" text field. This provides for much more flexibility and easily lends to dealing with the rest of the possibly required fields (i.e. gender, level_of_education, etc.) in case someone uses an SAML IdP that passes these in.
2. This is not currently based on https://github.com/edx/edx-platform/pull/15135 but will be as soon as that gets merged.

**Reviewers**
- [ ] @haikuginger 
- [ ] edX reviewer[s] TBD

- - -
**Settings**
```yaml
configuration_version: ent-423
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
  ENABLE_THIRD_PARTY_AUTH: true
SWAPFILE_SIZE: 4GB
```

edx-webhook commented:
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=15305"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

UmanShahzad commented:
Currenty WIP, should be ready soon.

bradenmacdonald commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -241,22 +241,31 @@ def get_register_form_data(cls, pipeline_kwargs):
+            attributes = pipeline_kwargs['response']['attributes']
+            for attr, val in current_provider.required_extra_attributes.iteritems():
+                if val:
+                    registration_form_data[attr] = attributes[val]
```
I'm not convinced that it adds much value to make this `required_extra_attributes` defined on the provider... Would there be any downside to simply writing

```python
registration_form_data['country'] = details.get('country', '')
```
unconditionally?

Or if it is conditional, I feel like it should be conditional on whether or not the Open edX instance has the "country" field enabled for this form - not whether or not the provider says it sometimes "requires" the extra country attribute.

Put another way: I can't think of a reason why we wouldn't want to pre-fill the country field if it's enabled and if the data is available from the provider, so I don't think we need to make that behavior optional at any level.

UmanShahzad replied:
@bradenmacdonald Well for one, it wouldn't be coming in `details` because that's something PSA fills for us and PSA isn't filling up country data.

In that case one would have to gather this data directly from `pipeline_kwargs['response']['attributes']` the way I attempt here above. But not all backends get their data in this way, so this wouldn't work as an unconditional, general functionality.

bradenmacdonald commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -379,6 +392,9 @@ class SAMLProviderConfig(ProviderConfig):
         help_text="URN of SAML attribute containing the user's email address[es]. Leave blank for default.")
+    attr_country = models.CharField(
+        max_length=128, blank=True, verbose_name="Email Attribute",
+        help_text="URN of SAML attribute containing the user's ISO-3166-1 Alpha-2 country code. Leave blank for default.")
```
This says "Leave blank for default.", but [unlike the other attributes](https://github.com/python-social-auth/social-core/blob/ace9733/social_core/backends/saml.py#L55-L66), I don't think that any default is implemented.

If there were a default, I think we would want it to be `urn:oid:2.5.4.6` - you could put that as an example in the help string - `e.g. "urn:oid:2.5.4.6" or leave blank to not expect a country attribute.`

UmanShahzad replied:
@bradenmacdonald Ah this fact slipped by me. I think that since the other attributes do actually default to something, this should too. We can define it as a constant at the top there.

UmanShahzad commented:
@bradenmacdonald Fixed the default issue -- decided to just actually default it the way the rest are defaulted.

Posted a hefty response to the other review, though -- I think the way I've done it is _okay_ for now (only really works for a few cases), but it's still a little more general (imo) than going for the country code directly, and period -- maybe I can rework the solution a bit but for the current ticket (aimed at SAML) it should do. That is, of course, I understood your concerns correctly.

haikuginger commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -225,7 +227,7 @@ def get_social_auth_uid(self, remote_id):
 
     @classmethod
-    def get_register_form_data(cls, pipeline_kwargs):
+    def get_register_form_data(cls, pipeline_kwargs, **kwargs):
```
Passing an instance as a keyword argument to a class method seems like a Bad Thing. If we need that data, make it an instance method (remove the `@classmethod` decorator and change the `cls` arg to `self`).

Unless I'm missing something here?

UmanShahzad replied:
I've tried this, but was getting failures in a bunch of tests which started to look awfully awkward to solve, so I tried to work around it with this technique. Going to take another look at this solution and get back to you.

haikuginger replied:
Thanks!

haikuginger commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -241,22 +243,31 @@ def get_register_form_data(cls, pipeline_kwargs):
+
+        # There are often nonstandard requirements for some registration flows,
+        # like the user's country code.
+        if current_provider.required_extra_attributes:
```
If we leave `current_provider` as an optional kwarg, here's where we get our first AttributeError when we don't pass it.

UmanShahzad replied:
Sorry, can fix this by defaulting to `cls` as what we pop from the `kwargs`.

haikuginger commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -379,6 +394,9 @@ class SAMLProviderConfig(ProviderConfig):
         max_length=128, blank=True, verbose_name="Email Attribute",
         help_text="URN of SAML attribute containing the user's email address[es]. Leave blank for default.")
+    attr_country = models.CharField(
+        max_length=128, blank=True, verbose_name="Email Attribute", default=OID_COUNTRY_NAME,
```
Attribute verbose name.

UmanShahzad replied:
Ah, jeez. Thanks.

haikuginger commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -277,6 +288,10 @@ def enabled_for_current_site(self):
         return self.enabled and self.site == Site.objects.get_current(get_current_request())
 
+    @property
+    def required_extra_attributes(self):
```
Rationale to do this as a property as opposed to a class variable?

UmanShahzad replied:
I suppose it's not a very good one :)

UmanShahzad replied:
@haikuginger Update: see comment below.

haikuginger commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -468,6 +487,10 @@ def get_config(self):
         return idp_class(self.idp_slug, **conf)
 
+    @property
+    def required_extra_attributes(self):
```
As above - why a property and not a variable if the value is static?

UmanShahzad replied:
Again, not a very good rationale -- I believe at first I was trying some weird solution that required me to use it this way, but I reworked it and no longer needed it but kept it as a property. I'm going to revert it to a class variable.

UmanShahzad replied:
@haikuginger I went back into it and realized the reason I needed this to be a property is because it depends on `self` for getting the value of `attr_country` (we need the URN).

haikuginger replied:
Ah, got it. I missed that.

haikuginger commented on openedx/core/djangoapps/user_api/views.py:
```diff
@@ -921,10 +923,11 @@ def _apply_third_party_auth_overrides(self, request, form_desc):
                     )
 
-                    for field_name in self.DEFAULT_FIELDS:
+                    for field_name in self.DEFAULT_FIELDS + self.EXTRA_TPA_FIELDS:
```
I'm not sure if this is the right construct to get this information. Why do we need to know that these fields are "TPA fields"? Why not just iterate through all the fields for the form, while being aware that we might not get a value?

UmanShahzad replied:
I thought it'd be maybe better to separate out those fields that TPA wants specifically, in addition to the defaults. I contemplated just doing `self.DEFAULT_FIELDS + self.EXTRA_FIELDS` but I suppose it's less code maintenance to just have it check every field.

mattdrayer commented:
Just a heads up that I would consider the merge deadline for this functionality more urgent than "not that urgent" -- as a matter of fact @bradenmacdonald @haikuginger @douglashall if we could possibly resolve this one today, that would be a very good thing imho, since it supports a client integration case.

UmanShahzad commented:
@mattdrayer OK, thanks for the heads up on this. I'm working on addressing the reviews ASAP.

UmanShahzad commented:
@haikuginger Thanks for your review -- I'll try to fix up things accordingly and get back to you ASAP as this ticket seems urgent according to @mattdrayer.

UmanShahzad commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -241,22 +243,31 @@ def get_register_form_data(cls, pipeline_kwargs):
+            attributes = pipeline_kwargs['response']['attributes']
+            for attr, val in current_provider.required_extra_attributes.iteritems():
+                if val:
+                    registration_form_data[attr] = attributes.get(val, '')
```
I'll remove the if check here in the next push.

UmanShahzad commented:
@haikuginger I believe I fixed / addressed the issues from the review, except I didn't have time to [do what I mentioned I'd do in comments](https://github.com/edx/edx-platform/pull/15305#discussion_r121927835).

Can we go for one more review on this?

UmanShahzad commented:
Nevermind my comment about the unit tests; it's already tested well enough in [`assert_register_response_in_pipeline_looks_correct`](https://github.com/edx/edx-platform/blob/master/common/djangoapps/third_party_auth/tests/specs/base.py#L254).

UmanShahzad commented:
@haikuginger The changes you made seem to make it work, now. :+1: 

Testing log below:

```
(Pdb) details.update({"country": "US"})
(Pdb) pp details
{'country': 'US',
 'email': None,
 'first_name': 'Me Myself',
 'fullname': 'Me Myself And I',
 'last_name': 'And I',
 'username': 'myself'}
(Pdb) cont
[16/Jun/2017 06:16:02] "POST /auth/complete/tpa-saml/ HTTP/1.1" 302 0
[16/Jun/2017 06:16:02] "GET /register HTTP/1.1" 200 58941
```

And it successfully defaulted the Country field to "United States of America":

![](http://i.imgur.com/GJGay4J.png)

I'll fixup the test failures and write a test for this new feature asap.

haikuginger commented on openedx/core/djangoapps/user_api/tests/test_helpers.py:
```diff
@@ -147,6 +147,38 @@ def test_invalid_restriction(self):
+        }
+        desc.override_field_properties(
+            field["name"],
+            options=[
```
Instead of setting `options` in the `desc.override_field_properties` call, set `default='PK'`, and then pass the options (without default parameters) to the field constructor as you normally would - that's the path we're using.

In other words, this chunk should look like this:

```python
desc.override_field_properties(
    field['name'],
    default='PK'
)
```

UmanShahzad replied:
I see, okay.

haikuginger commented on openedx/core/djangoapps/user_api/helpers.py:
```diff
@@ -190,9 +190,10 @@ def add_field(
 
             options (list): For "select" fields, a list of tuples
-                (value, display_name) representing the options available to
+                (value, display_name, default) representing the options available to
```
We're not using this `default` parameter; please remove it. What we're doing is taking the normal `default` kwarg _for the field as a whole_; then, for each value we add to the select field, we're _marking_ it as a default if it matches the arg we received.

UmanShahzad replied:
Hmm, I see. It's funny I changed this comment when in the test I wrote I didn't even do it this way :X

haikuginger commented on openedx/core/djangoapps/user_api/tests/test_views.py:
```diff
@@ -1030,15 +1010,15 @@ def test_register_form_level_of_education(self):
+                    {"value": "jhs", "name": "Junior secondary/junior high/middle school", "default": False},
+                    {"value": "el", "name": "Elementary/primary school", "default": False},
+                    {"value": "none", "name": "No formal education", "default": False},
+                    {"value": "other", "name": "Other education", "default": False},
```
Can you add a test to this suite that should have an overridden default value for a select-type field?

UmanShahzad replied:
Sure.

UmanShahzad commented:
@haikuginger Ah, okay that makes sense.

UmanShahzad commented:
Tests are passing -- YouTube was having some problems so ignore the single irrelevant test failure.

UmanShahzad commented:
@mattdrayer I'm currently continuing to write tests for the user API views to further build confidence, going to be pushing that soon.

The testing I did above was for registering using TestShib as an SAML IdP -- is that what you mean?

haikuginger commented:
@UmanShahzad, @mattdrayer means for the case where the SAMLProviderConfiguration has "Skip registration form" set. Note that for that to work, you'll need to configure the email field to use the URN that TestShib is using.

@mattdrayer, I did verify that case, but let's have @UmanShahzad verify it as well.

bradenmacdonald commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -264,6 +262,16 @@ def get_register_form_data(self, pipeline_kwargs):
+        # because the fields that are actually rendered are not based on this
+        # list, only those values that map to a valid registration form field
+        # will actually be sent to the form as default values.
+        for blacklisted_field in REGISTRATION_FORM_FIELD_BLACKLIST:
```
Nit (feel free to ignore since I don't want to hold this up): Since `REGISTRATION_FORM_FIELD_BLACKLIST` is not used anywhere else, IMHO it is much easier to follow the logic of this code if we just put the value `['name', 'username']` inline here, rather than making the user read something elsewhere in the file to understand what's going on.

Even better, just put `details.pop('name', None)  # Ensure we don't overwrite the name key later` and `details.pop('username', None)` inline above, each on the line following the respective `registration_form_data['...'] = `

bradenmacdonald commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -420,10 +417,13 @@ class SAMLProviderConfig(ProviderConfig):
-            'use of the information stored in this field for configuration.'
+            'The tpa-saml backend supports {"requiredEntitlements": ["urn:..."]}, '
+            'which can be used to require the presence of a specific eduPersonEntitlement, '
+            'and {"extra_field_definitions": [{"name": "...", "urn": "..."},...]}, which can be '
```
Nit: e.g. `{"country": "urn:oid:1.2.3.4", "date_of_birth": "urn:oid:365.86400.0"}` would be a simpler data format.

mattdrayer replied:
Now that I've started using this feature in production for one of our enterprise customers, I agree that @bradenmacdonald's comment here would be nice to have.  Also sometimes the URN value is not provided in the `<saml:Attribute>` element  -- for example, in my case the "Name" attribute is "countryCode":  

```
<saml:Attribute Name="countryCode" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
    <saml:AttributeValue xsi:type="xs:string">in</saml:AttributeValue>
</saml:Attribute>
```

I'm assuming that it's fine to specify "countryCode" for the "urn" value in this case, but we might consider another iteration on this at some point.

UmanShahzad commented:
I'm going to test the auto-registration scenario next.

UmanShahzad commented:
@mattdrayer @haikuginger OK, I can confirm that auto-registration keeps the country value when input into attributes from a breakpoint.

![](http://i.imgur.com/f0PTTZR.png)

UmanShahzad commented:
@brittneyexline Thanks! Does this sound good for a squash then?

brittneyexline commented:
@UmanShahzad, yes, please go ahead and squash

UmanShahzad commented:
@brittneyexline Ok, I've done it, but not sure why GitHub doesn't show me it as having been updated on this PR. The branch https://github.com/open-craft/edx-platform/tree/uman/OC-2682 is updated, though.

UmanShahzad commented:
jenkins run bokchoy

UmanShahzad commented:
jenkins run js

UmanShahzad commented:
jenkins run python
